webui/react/src/pages/WorkspaceDetails/WorkspaceProjects.tsx

```diff
@@ -70,48 +72,31 @@ const WorkspaceProjects: React.FC<Props> = ({ workspace, id, pageRef }) => {
   const config = useMemo(() => configForWorkspace(id), [id]);
   const { settings, updateSettings } = useSettings<WorkspaceDetailsSettings>(config);
 
-  const fetchProjects = useCallback(async () => {
-    if (!settings) return;
+  const loadableProjects: Loadable<List<Project>> = useObservable(
+    projectStore.getProjectsByWorkspace(id),
+  );
 
-    try {
-      const response = await getWorkspaceProjects(
-        {
-          archived: workspace?.archived ? undefined : settings.archived ? undefined : false,
-          id,
-          limit: settings.view === GridListView.Grid ? 0 : settings.tableLimit,
-          name: settings.name,
-          offset: settings.view === GridListView.Grid ? 0 : settings.tableOffset,
-          orderBy: settings.sortDesc ? 'ORDER_BY_DESC' : 'ORDER_BY_ASC',
-          sortBy: validateDetApiEnum(V1GetWorkspaceProjectsRequestSortBy, settings.sortKey),
-          users: settings.user,
-        },
-        { signal: canceler.signal },
-      );
-      setTotal(response.pagination.total ?? 0);
-      setProjects((prev) => {
-        if (_.isEqual(prev, response.projects)) return prev;
-        return response.projects;
-      });
-    } catch (e) {
-      handleError(e, { publicSubject: 'Unable to fetch projects.' });
-    } finally {
-      setIsLoading(false);
-    }
-  }, [canceler.signal, id, workspace, settings]);
+  useEffect(() => {
+    Loadable.match(loadableProjects, {
+      _: () => {},
+      Loaded: (data) => {
+        setProjects(data);
+        setIsLoading(false);
+        streamStore.emit(new ProjectSpec([id]));
+      },
+    });
+  }, [loadableProjects, id]);
 
   useEffect(() => {
-    setIsLoading(true);
-    fetchProjects().then(() => setIsLoading(false));
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
+    setTotal(projects.size);
+  }, [projects]);
```

Comment:
avoid directly setting state in effects -- that causes the render to have to rerun when the dependencies change.
Instead, derive the variable from the dependent inline or use `useMemo` if the calculation is expensive.

i.e.
```tsx
//project list, loading, total are derived from project observable
const project = useMemo(() => loadableProjects.getOrElse(List(), [loadableProjects]);
const loading = useMemo(() => loadableProjects.isNotLoaded, [loadableProjects]);
const total = useMemo(() => loadableProjects.map((p) => p.size).getOrElse(0), [loadableProjects]);
useEffect(() => {
  loadableProjects.forEach(() => streamStore.emit(new ProjectSpec([id])));
}, [loadableProjects, id])
```

if you're worried about DRY, you could combine them and use destructuring in the memo:
```tsx
const [projects, loading, total] = useMemo(() => {
  return loadableProjects.map((p): [Project[], boolean, number] => [p.toJS(), false, p.size])).getOrElse([[], true, 0]);
}, [loadableProjects])
useEffect(() => {
  loadableProjects.forEach(() => streamStore.emit(new ProjectSpec([id])));
}, [loadableProjects, id])
```